### ISSUE TITLE
Introduce HTTP Filters

### ISSUE THREAD

**Issue by chamil321** [ballerina-platform/ballerina-standard-library#692]
$Title
Subtask of #576

Task breakdown
> [Phase 1 - Request Interceptors](https://github.com/ballerina-platform/ballerina-standard-library/issues/2062)
> [Phase 2 - Response Interceptors](https://github.com/ballerina-platform/ballerina-standard-library/issues/2684)

**Comment by shafreenAnfar** [comment]
Also, I think instead of the name Filters, I think we can use Interceptors or Middleware

**Comment by shafreenAnfar** [comment]
I think the better name for `FilterContext` is `RequestContext`. Because the `RequestContext` is instigated and scoped by the request.

**Comment by ldclakmal** [comment]
+1 for the suggested proposal which aligns with the Swan Lake service design. Below are the summarized actions that a user can do on top of each component as per my understanding.

**`RequestFilter`**
- call `ctx.next()` and return `()` --> call the next filter in chain
- return `error` --> call to the `RequestErrorFilter`
- return `()` --> the request will be left hanging (which is okay since the user does not adhere to the concept)
- use `http:Caller` and respond --> ?

**`ResposeFilter`**
- call `ctx.next()` and return `()` --> call the next filter in chain
- return `error` --> call to the `ResponseErrorFilter`
- return `()` --> the request will be left hanging (which is okay since the user does not adhere to the concept)

**`RequestErrorFilter`**
- call `ctx.next()` and return `()` --> ?
- return `error` --> ?
- return `()` --> ?
- use `http:Caller` and respond --> ?

**`ResponseErrorFilter`**
- call `ctx.next()` and return `()` --> ?
- return `error` --> ?
- return `()` --> ?

Few things the clarify,
- Related to the `RequestFilter`
  -  What is the expected behaviour when the user use `http:Caller` and respond?
 
- Related to the `RequestErrorFilter` or `ResponseErrorFilter`
  - What is the expected behaviour when the `[Request|Response]ErrorFilter` returns an error?
  - Assuming returning `()` is the success case, does the user always need to use `http:Caller` and respond before returning `()`.
  - What is the expected behaviour if a user call `ctx.next()`?
  - Do we really need `[Request|Response]ErrorFilter`? Can't we handle the error scenario from `[Request|Response]Filter` itself? If an error is returned without handling, the same behaviour of HTTP resource returning an error.

**Comment by shafreenAnfar** [comment]
> What is the expected behaviour when the user use http:Caller and respond?

Which basically means user has terminated the request/response cycle. But that has nothing to do with Interceptors. As long as user has called `ctx.next()` execution will continue. But the request will never get dispatched to the particular resource. 

> What is the expected behaviour when the [Request|Response]ErrorFilter returns an error?

Framework will simply log the error and continue. 

> What is the expected behaviour if a user call `ctx.next()` ?

Nothing will happen as there is nothing to continue. 

> Do we really need [Request|Response]ErrorFilter? 

Yes. 

These errors are unhandled errors which could be either expected or unexpected. In either case, user is given the opportunity to handler these errors the way they want. They can simply send back a response, log, collect matrices, etc, and all the error handling code will be at one place, in other words there is no need to repeat.

If we send back a response user can't do none of the above but settle for 500 internal server response which is too restrictive. Furthermore, if we not chose  to send 500 internal server error, then the only other option is to log the error and continue the pipeline which is again too restrictive. It also breaks the consistency because it results in two ways to continue the pipeline. 

Lastly, similar error handling options can be found in standard pipeline implementations such as Netty and Express Middleware. 

Therefore, I think having error handlers not only makes the design much cleaner but also gives the user more freedom.

**Comment by shafreenAnfar** [comment]
Actually, I think we can further improve the pipeline as per @ldclakmal's suggestion. 

If the user hasn't added an error handler, the originally proposed design suggest we have a default handler which basically logs the error and do nothing. 

Instead, if the user hasn't added an error handler, returning an error results in 500 internal server error, which also means the request/response cycle get terminated at that point. If the user has added an error handler, then the execution jumps to the error handler as originally proposed. Then if the user returns an error within the error handler it also translates to 500 internal server error similar to any interceptor. Besides, error-handler is just another interceptor.

This applies to both Request and Response paths.

**Comment by shafreenAnfar** [comment]
Also, I think we can let users add error handlers to any place in the pipeline. Instead of fixing them to a predefined location. In that case, execution jumps to the nearest error handler in the pipeline. 

Lastly, to keep things simple, we can say that, `ctx.next()` must be invoked in the interceptor that is there before the actual service. That way the final service also becomes part of the pipeline.

**Comment by shafreenAnfar** [comment]
I think, we can further simplify the design by allowing to return `RequestFilter` and `ResponseFilter` in the respective method signatures. Calling `ctx.next()` would simply return the reference of the next interceptor in the chain which then get executed by the framework. 

The advantage of this is,
- We can stick to all the rules we already have for resources such as returning to respond, using the `Caller` etc.
- Eliminate the confusion as to what happens when `ctx.next()` is invoked in the middle of a function logic. 

Following is an example for `RequestFilter` based on the newly proposed design. 

```Ballerina
service class RequestFilter {
   *http:RequestFilter;
 
   remote function 'default [string… path](http:FilterContext ctx, http:Caller caller,
                       http:Request req) returns RequestFilter|error? {
      // do some work
       return ctx.next();
   }
}
```
Similarly `ResponseFilter` would be as follows,

```Ballerina
service class ResponseFilter {
   *http:ResponseFilter;
 
   remote function filterResponse(http:FilterContext ctx, http:Response res) returns ResponseFilter|error? {
        // do some work
       return ctx.next();
   }
}
```
Lastly, `RequestErrorFilter` and `ResponseErrorFilter` would be as follows,

```ballerina
service class RequestErrorFilter {
   *http:RequestErrorFilter;
 
   remote function 'default [string… path](http:FilterContext ctx, http:Caller caller,
                       http:Request req, error err) returns RequestFilter|error? {
       // deal with the error
   }
}
```
```Ballerina
service class ResponseErrorFilter {
   *http:ResponseErrorFilter;
 
   remote function filterResponseError(http:FilterContext ctx,
                       http:Response res, error err) returns ResponseFilter|error? {
       // deal with the error
       return ctx.next()
   }
}
```
> **Note** : We can only additionally return `RequestFilter` and `ResponseFilter`.

@ldclakmal any thoughts ?

**Comment by ldclakmal** [comment]
As a summary of my understanding:
- The pipeline can have one or more filters.
- A filter can be one of the `RequestFilter`, `ResponseFilter`, `RequestErrorFilter` or `ResponseErrorFilter`.
- There is no such pre-defined order, it is up to the user. During the time of execution, the pipeline will get executed in the configured order.
- For the success case, each filter should return `ctx.next()`.
- For the error case, each filter will return the `error` and the execution will immediately jump to the nearest `[Request|Response]ErrorFilter` based on the type of calling filter.
- If there is no such `[Request|Response]ErrorFilter` in the pipeline, it would result in a 500 response.
- Since the `[Request|Response]ErrorFilter` is treated as another filter, the same concepts are applied for those as well.
- The only difference would be returning an error from the `[Request|Response]ErrorFilter` would result in a 500 error response.
- Returning `()` from any type of the filters without returning the `ctx.next()` would result in a 500 error response which is 100% similar to what is happening at the HTTP resource when the `http:Caller` is available.

These address all the concerns I raised above and address the possible use cases I can think of. Hence I am +1 for this design.

NOTE: As suggested above, I am +1 for updating all the `Filter` word usages with the `Interceptor` word in all objects, records, variables, parameters etc.

**Comment by chamil321** [comment]
Very informative proposal and has covered all aspects.

The word request filter itself explains that it is used to perform some action to a selected group of request. The word interceptor means to perform some action before executing the business logic. The motivation of the proposal mentions facts related to interceptors. Hence +1 for the word `Interceptor`

But I see there is an involvement of the filter concepts too. One certain occasion is the dispatching concepts which brought into Interceptors. My suggestions based on the motivation being the ability to execute some common logic for all the inbound requests and outbound responses is to **avoid resource dispatching concepts from the RequestInterceptor**. 

IINM RequestFilter is not a network endpoint but the Resource method. Since the motivation is to execute common logic, the HTTP verb and path separations should not encounter IMO. If we are to execute some logic on path `foo`, then that logic should be in the `foo` resource method. Also binding some data to typed payload could not be common to all the resources of the service. 

Therefore suggestion is to make both Request and Response filters consistent and have remote methods of `filterRequest`/`filterResponse`

**Comment by shafreenAnfar** [comment]
Interesting, however, we shouldn't restrict the users by the design. Also, request and response are two different things they have different properties, therefore, there is no point int trying make request and response interceptors consistent. 

Moreover, when it comes to the request path, we must strive to keep things consistent. That is what the design does. There is very little difference between the RequestInterceptor/RequestErrorInterceptor and the final service. In other words those are consistent with one another. They are all just services with similar properties attached to a pipeline. 

Lastly, as @chamil321 mentioned, within the pipeline there could be filtering when it comes to engaging the interceptors. I think that gives users more freedom to come up with better solutions for their problem. Besides, all the popular frameworks such as gorilla/mux, express and springboot support this.

**Comment by chamil321** [comment]
Clarification on the resource dispatching logic implementation.

#### Engagement level : Listener
Intercepter service basePath default to `/`
Resource name should contain both the basePath and resource path

```ballerina
service class RequestInterceptor {
   *RequestInterceptor;
 
   resource function 'default foo/bar(http:FilterContext ctx, http:Caller caller,
                       http:Request req) returns RequestFilter|error? {
      // do some work
       return ctx.next();
   }
}

RequestInterceptor requestInterceptor = new();

listener http:Listener echoListener = new http:Listener(9090, config = {filters: [requestInterceptor]});

service /foo on echoListener {
    resource function 'default bar() returns json {
        return { "message":"response from service" };
    }
}

```
Having only the `bar` as the resource  name of interceptor service does not specifically helps to find the exact match. Because there can be many resources with the name bar in other services as well. 

#### Engagement level : Service

Service basePath will be used.

@shafreenAnfar Please add thoughts if you think otherwise

**Comment by chamil321** [comment]
In order to distinguish the Request/Response/Error Interceptors, the type should be distinct as compiler cannot uniquely identify the service type yet.  
Related : https://github.com/ballerina-platform/ballerina-lang/issues/33595

```ballerina
# The HTTP request interceptor service object  
public type RequestInterceptor distinct service object {

};

# The HTTP request error interceptor service object  
public type RequestErrorInterceptor distinct service object {

};
```

--

If the Request/Error interceptor has responded using the caller and called `ctx.next()`, then Response intercepter are invoked before continuing to the next next interceptor. It is assumed that the Response interceptors are called following the respond() method sequentially. 

```ballerina
resource function 'default [string...path](http:RequestContext ctx, http:Caller caller, http:Request request) returns http:InterceptorReturnValues {
   http:Response res = new();
   res.setHeader("last-interceptor", "request-interceptor-caller-repond");
   res.setTextPayload("Response from caller inside interceptor");
   check caller->respond(res);
   return ctx.next();
}
```

In a nutshell, both Response and Request interceptor are not executed in parallel in this instance.

**Comment by shafreenAnfar** [comment]
> In a nutshell, both Response and Request interceptor are not executed in parallel in this instance.

Yeah, sequential execution keeps things simple as `RequestContext` will be shared with both types of interceptors.

**Comment by shafreenAnfar** [comment]
I think we have done a slight mistake when it comes to `RequestContext`'s `add` API. The name of the API should be `set` as calling the method with the same `key` value results in replacing the existing value.

**Comment by TharmiganK** [comment]
When we have interceptors configured in Listener level and Service level, there are two options
1. Execute both of them
2. Overwrite the listener level interceptor pipeline with the service level one

Option 1 can be used if the user want to execute some common logic for all the services. But for this we need the order of execution : do we execute listener level or service level first?

Option 2 is simple and less complicated than Option 1. Also if the user wants to execute some common logic, the user can always add the common interceptor to every service level interceptors

@shafreenAnfar @chamil321 WDYT?

**Comment by shafreenAnfar** [comment]
I think the third option is to limit users to only write Interceptors with default path (`[string... path]`) at listener level. That way if users want to engage interceptors only for a particular service they must have it at service level. This leads to more readable code compared to previous two options.

**Comment by TharmiganK** [comment]
> I think the third option is to limit users to only write Interceptors with default path (`[string... path]`) at listener level. That way if users want to engage interceptors only for a particular service they must have it at service level. This leads to more readable code compared to previous two options.

+1 for this option

In this case, when we have interceptors engaged in both listener and service level, the interceptor pipeline flow will be :

`ListenerLevel:RequestInterceptors` >> `ServiceLevel:RequestInterceptors` >> `Target Service` >> `ServiceLevel:ResponseInterceptors` >> `ListenerLevel:ResponseInterceptors`

**Comment by shafreenAnfar** [comment]
I think supporting data binding for request interceptors could cause trouble. For instance,

1. First request interceptor binds to a record and changes a field in the record 
2. Second request interceptor binds to JSON 

In such cases, it opens up a lot of issues. Especially because we are relying on the reference of the request to pass changes from one to another. Therefore, we shouldn't support data-binding for request interceptors.

**Comment by chamil321** [comment]
@shafreenAnfar Let me explain our thought process when we designing the data-binding support for the interceptors.

The interceptors executed in an order in a pipeline. The change are done to the common arguments(request, payload, header) should be propagated to the following interceptors and finally to the HTTP service. 

When it comes to the data binding, if the there are multiple interceptors with `@http:Payload` annotation, the first data binding required interceptor will build the data source of the request and set it in the request entity.

The subsequent interceptors will check the build data-source first and proceed. 

If the accessed payload is changed/altered inside an interceptor, and set it back again as the payload of the request, then that changed/altered payload is visible to the next interceptor. In case of not setting back to the request will not reflect that change in the next interceptor.

Since the request is consist of headers and payload, the changes are done to both headers and payload is reflected in next interceptor when that are set back to the request after the change. 

> I think supporting data binding for request interceptors could cause trouble. For instance,
> 
> 1. First request interceptor binds to a record and changes a field in the record
> 2. Second request interceptor binds to JSON
> 
If we set back the payload to the request, the change is propagated. Otherwise, change is only applicable within the first interceptor

**Comment by TharmiganK** [comment]
For Example : 
```ballerina
import ballerina/http;
import ballerina/io;

type Person record {|
    string name;
    int age;
|};

service class RequestInterceptor1 {
    *http:RequestInterceptor;

    resource function 'default [string... path](http:RequestContext ctx, @http:Payload Person person, 
            http:Request req) returns http:NextService|error? {
        io:println(person.toString());
        person.name = "Tharmi";
        req.setPayload(person);
        return ctx.next();
    }
}

service class RequestInterceptor2 {
    *http:RequestInterceptor;

    resource function 'default [string... path](http:RequestContext ctx, @http:Payload json person, 
            http:Request req) returns http:NextService|error? {
        io:println(person.toString());
        return ctx.next();
    }
}

RequestInterceptor1 requestInterceptor1 = new;
RequestInterceptor2 requestInterceptor2 = new;

listener http:Listener interceptorListener = new http:Listener(9090, config = {
    interceptors: [requestInterceptor1, requestInterceptor2] 
});

service / on interceptorListener {
    resource function 'default . () returns string {
        return "HelloWorld";
    }
}
```
If we send the following curl request : `curl "http://localhost:9090" -d '{"name":"Krishna","age":25}' -H 'Content-Type: application/json'` we get the following output : 
```
{"name":"Krishna","age":25} - From Interceptor 1
{"name":"Tharmi","age":25}  - From Interceptor 2
{"name":"Tharmi","age":25}  - From Main resource
```
If we do not set the payload in the first interceptor, then the payload will not change :
```
{"name":"Krishna","age":25} - From Interceptor 1
{"name":"Krishna","age":25} - From Interceptor 2
{"name":"Krishna","age":25} - From Main resource
```

**Comment by shafreenAnfar** [comment]
Got it. I think this design makes sense.

**Comment by shafreenAnfar** [comment]
Btw, once the proposal is implemented, we need to check https://github.com/ballerina-platform/ballerina-standard-library/issues/2681

**Comment by TharmiganK** [comment]
Clarification on response interceptor flow :

**Case 1 :** 
```ballerina
service class ResponseInterceptor1 {
    *http:ResponseInterceptor;

    remote function interceptResponse(http:Response res) {
        res.setHeader("response-interceptor", "interceptor-1");
    }
}
```
Response interceptor without responding or calling `RequestCtx.next()` :

> Returning the response at this point and stop the interceptor pipeline

**Case 2 :** 
```ballerina
service class ResponseInterceptor2 {
    *http:ResponseInterceptor;

    remote function interceptResponse(http:Response res, http:Caller caller) returns error? {
        res.setHeader("response-interceptor", "interceptor-2");
        return caller->respond(res);
    }
}
```
Response interceptor with caller respond  but not calling `RequestCtx.next()` :

> Returning the response at this point and stop the interceptor pipeline. Similar to Case 1, but caller can respond any values where as in Case 1 we only return the referenced response object.

**Case 3 :** 
```ballerina
service class ResponseInterceptor3 {
    *http:ResponseInterceptor;

    remote function interceptResponse(http:Response res, http:Caller caller, http:RequestCtx ctx)
            returns http:NextService|error? {
        res.setHeader("response-interceptor", "interceptor-3");
        check caller->respond(res);
        return ctx.next();
    }
}
```
Response interceptor with caller respond and calling `RequestCtx.next()` :

> Returning the response at this point and then proceed with the next response interceptors which will ends up in `already responded error` or `response is already used error` which is similar to request interceptors.

@shafreenAnfar @chamil321 Please add your thoughts if you think otherwise

**Comment by chamil321** [comment]
@shafreenAnfar above cases are missed in the spec. Please raise if there are any concerns. Otherwise we will go ahead with above flow and update the spec

**Comment by shafreenAnfar** [comment]
Firstly, it is nice to see you guys have come up with the solution along with the problem.

How I see the problem is responding always trigger the next immediate response interceptor. Responding can be done in two different ways either via the caller or via the `return` statement. This is true for both request interceptors as well as response interceptors. Therefore, I believe the the right thing to do in **case 2** and **case 3** is to execute the next immediate response interceptor. 

`next()` always triggers the next immediate interceptor. In **case 3** it will just try to re-execute the same interceptors. Therefore, in  **case 3** calling `next()` is redundant. 

I think the design wise staying consistent is important.

**Comment by chamil321** [comment]
> How I see the problem is responding always trigger the next immediate response interceptor. Responding can be done in two different ways either via the caller or via the `return` statement. This is true for both request interceptors as well as response interceptors. Therefore, I believe the the right thing to do in **case 2** and **case 3** is to execute the next immediate response interceptor.
 
Intuitively, how I felt was when the `caller->respond()` is called, the flow is stopped and respond to the caller. If there are response interceptors, they will get executed. If not, just send the response. `ctx.next()` is more like calling the next service IMO

1. Respond in a request interceptors : Stop the request interceptor and main service execution, execute the response flow
2. Respond in a response interceptors : Stop the response interceptors, send out the response.

However the issue is, there is noway to propagate new response object/value to the next response interceptor. For that, currently `caller->respond()` can be used.

**Comment by shafreenAnfar** [comment]
Yeah, lets return and error in this case.

**Comment by shafreenAnfar** [comment]
Closing the issue as the implementation is done.
